### PR TITLE
Switch to v2 object split scheme

### PR DESCRIPTION
**Pull request by smallhive**

### CODE REVIEW

File: internal/neofs/tree.go [hunk 1]
```diff
@@ -1320,7 +1334,6 @@ func metaFromMultipart(info *data.MultipartInfo, fileName string) map[string]str
 	info.Meta[uploadIDKV] = info.UploadID
```
Comment (roman-khimov) [inline internal/neofs/tree.go:1334]: Can the first object ID be used for UploadID?

File: api/layer/tree_service.go [hunk 2]
```diff
@@ -74,11 +74,10 @@ type TreeService interface {
 	// If object id to remove is not found returns ErrNoNodeToRemove error.
 	AddPart(ctx context.Context, bktInfo *data.BucketInfo, multipartNodeID uint64, info *data.PartInfo) (oldObjIDToDelete oid.ID, err error)
 	GetParts(ctx context.Context, bktInfo *data.BucketInfo, multipartNodeID uint64) ([]*data.PartInfo, error)
-	// GetLastPart returns the latest uploaded part.
-	//
+	// GetPartByNumber returns the part by number. If part was uploaded few times the newest one will be returned.
 	// Return errors:
 	//   - [ErrPartListIsEmpty] if there is no parts in the upload id.
-	GetLastPart(ctx context.Context, bktInfo *data.BucketInfo, multipartNodeID uint64) (*data.PartInfo, error)
+	GetPartByNumber(ctx context.Context, bktInfo *data.BucketInfo, multipartNodeID uint64, number int) (*data.PartInfo, error)
```
Comment (roman-khimov) [inline api/layer/tree_service.go:80]: Why this change? I don't see you requesting anything but the last part.

File: api/layer/neofs_mock.go [hunk 3]
```diff
@@ -270,19 +270,13 @@ func (t *TestNeoFS) CreateObject(_ context.Context, prm PrmObjectCreate) (oid.ID
...
 	t.currentEpoch++
 
-	if prm.Multipart != nil && prm.Multipart.SplitID != "" {
-		var split object.SplitID
-		if err := split.Parse(prm.Multipart.SplitID); err != nil {
-			return oid.ID{}, fmt.Errorf("split parse: %w", err)
-		}
-		obj.SetSplitID(&split)
-
+	if prm.Multipart != nil {
```
Comment (carpawell) [inline api/layer/neofs_mock.go:273]: so the whole `if` works right only because we pray it is called right? i mean a link object should never have a previous one sert, a regular object's payload should never be rewritten, etc

File: api/layer/tree_mock.go [hunk 4]
```diff
@@ -390,7 +390,18 @@ func (t *TreeServiceMock) GetLastPart(ctx context.Context, bktInfo *data.BucketI
 		return 1
 	})
 
-	return parts[len(parts)-1], nil
+	var pi *data.PartInfo
+	for _, part := range parts {
```
Comment (carpawell) [inline api/layer/tree_mock.go:394]: aren't parts sorted?

File: api/layer/tree_service.go (hunk 2 above)
Comment (carpawell) [reply api/layer/tree_service.go:80]: but how is it related to V1 -> V2 move?

File: api/layer/neofs_mock.go [hunk 5]
```diff
@@ -293,6 +293,39 @@ func (t *TestNeoFS) CreateObject(_ context.Context, prm PrmObjectCreate) (oid.ID
...
+					payload = append(payload, partialObject.Payload()...)
+				}
+			}
+
+			pid, isSet := prm.Multipart.HeaderObject.ID()
+			if !isSet {
+				return oid.ID{}, errors.New("HeaderObject id is not set")
+			}
+
+			realHeaderObj := object.New()
```
Comment (carpawell) [inline api/layer/neofs_mock.go:314]: can you, please, explain why we do think in tests in `CreateObject`?

File: api/layer/tree_service.go (hunk 2 above)
Comment (carpawell) [reply api/layer/tree_service.go:80]: it was expected to simplify the process. can you, please, describe a little why it is required to _now_ and was not needed _then_?

File: api/layer/tree_mock.go (hunk 4 above)
Comment (carpawell) [reply api/layer/tree_mock.go:394]: why not try to get part directly via slice index? is there any cases when N-th part is not N-th in slice?

File: cmd/s3-gw/app_settings.go [hunk 6]
```diff
@@ -413,6 +413,8 @@ func newLogger(v *viper.Viper) *Logger {
 		panic(fmt.Sprintf("build zap logger instance: %v", err))
 	}
 
+	zap.ReplaceGlobals(l)
```
[the remark this answers is not in the document]
Comment (carpawell) [reply cmd/s3-gw/app_settings.go:416]: i would say globals are not good in general, also, i think we do not use it in any of our projects

File: api/layer/tree_mock.go (hunk 4 above)
Comment (smallhive) [reply api/layer/tree_mock.go:394]: There may be more than one part with the same index. They are sorted by `ServerCreated` field. We should find the latest/actual one.

Comment (roman-khimov) [reply api/layer/tree_mock.go:394]: But why? Even when you reupload, you invalidate the old one, but you still have only one proper current one.

File: api/handler/multipart_upload.go [hunk 7]
```diff
@@ -648,12 +649,12 @@ func encodeListMultipartUploadsToResponse(info *layer.ListMultipartUploadsInfo,
 		m := MultipartUpload{
 			Initiated: u.Created.UTC().Format(time.RFC3339),
 			Initiator: Initiator{
-				ID:          u.Owner.String(),
+				ID:          hex.EncodeToString(u.OwnerPubKey.Bytes()),
```
Comment (roman-khimov) [inline api/handler/multipart_upload.go:652]: StringCompressed()

File: api/handler/multipart_upload.go [hunk 8]
```diff
@@ -648,12 +649,12 @@ func encodeListMultipartUploadsToResponse(info *layer.ListMultipartUploadsInfo,
...
 				DisplayName: u.Owner.String(),
 			},
 			Key: u.Key,
 			Owner: Owner{
-				ID:          u.Owner.String(),
+				ID:          hex.EncodeToString(u.OwnerPubKey.Bytes()),
```
Comment (roman-khimov) [inline api/handler/multipart_upload.go:657]: And everywhere below

File: internal/neofs/tree.go [hunk 9]
```diff
@@ -226,6 +228,15 @@ func newMultipartInfo(node NodeResponse) (*data.MultipartInfo, error) {
 			}
 		case ownerKV:
 			_ = multipartInfo.Owner.DecodeString(string(kv.GetValue()))
+		case ownerPubKeyKV:
+			payload, err := hex.DecodeString(string(kv.GetValue()))
+			if err != nil {
+				return nil, fmt.Errorf("get pub key: %w", err)
+			}
+
+			if err = multipartInfo.OwnerPubKey.DecodeBytes(payload); err != nil {
```
Comment (roman-khimov) [inline internal/neofs/tree.go:237]: https://pkg.go.dev/github.com/nspcc-dev/neo-go/pkg/crypto/keys#NewPublicKeyFromString

File: internal/neofs/tree.go (hunk 1 above)
Comment (roman-khimov) [reply internal/neofs/tree.go:1345]: So you can upload the first object (even with 0 payload!) and use its hash, right?

File: api/handler/object_list.go [hunk 10]
```diff
@@ -199,6 +200,15 @@ func fillContents(src []*data.ObjectInfo, encode string, fetchOwner bool) []Obje
 			ETag:         obj.HashSum,
 		}
 
+		if size, ok := obj.Headers[layer.AttributeDecryptedSize]; ok {
+			sz, err := strconv.ParseInt(size, 10, 64)
+			if err != nil {
+				zap.L().Warn("Failed to parse size from header", zap.String("header_value", size), zap.Error(err))
```
Comment (roman-khimov) [inline api/handler/object_list.go:206]: It's an error. It should be returned in some way.

File: internal/neofs/tree.go [hunk 11]
```diff
@@ -1022,7 +1025,18 @@ func (c *TreeClient) GetLastPart(ctx context.Context, bktInfo *data.BucketInfo,
 		return 1
 	})
 
-	return parts[len(parts)-1], nil
+	var pi *data.PartInfo
+	for _, part := range parts {
```
Comment (roman-khimov) [inline internal/neofs/tree.go:1040]: Can be a lot of parts.

File: api/layer/tree_mock.go (hunk 4 above)
Comment (smallhive) [reply api/layer/tree_mock.go:394]: What do you mean invalidate? There is no such functionality. Only [add and get](https://github.com/nspcc-dev/neofs-s3-gw/blob/06acb581f7b68f0b7412e0dd340fb25f75d7aef1/api/layer/tree_service.go#L82). Any reuploaded parts presented in the list.

File: internal/neofs/tree.go (hunk 11 above)
Comment (smallhive) [reply internal/neofs/tree.go:1038]: Yes, it can. Right now there is no mechanism to reduce their amount. We may have separate issue for this

File: internal/neofs/tree.go (hunk 1 above)
Comment (smallhive) [reply internal/neofs/tree.go:1343]: No, any part must have at least 5Mb

File: internal/neofs/tree.go (hunk 9 above)
Comment (smallhive) [reply internal/neofs/tree.go]: Updated

File: api/handler/object_list.go (hunk 10 above)
Comment (smallhive) [reply api/handler/object_list.go]: Returned

File: cmd/s3-gw/app_settings.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (smallhive) [reply cmd/s3-gw/app_settings.go]: Well, It is not clear for me why you don't like global loggers, but 2vs1 - removed

File: api/layer/tree_mock.go (hunk 4 above)
Comment (smallhive) [reply api/layer/tree_mock.go:394]: We may try to filter out duplicates on the level down and try to have there is only valid part list

Comment (roman-khimov) [reply api/layer/tree_mock.go:394]: OK, seems to be unavoidable now.

File: internal/neofs/tree.go (hunk 1 above)
Comment (roman-khimov) [reply internal/neofs/tree.go:1343]: I mean NeoFS side. The first object can be created during upload init.

Comment (smallhive) [reply internal/neofs/tree.go:1343]: Well, in this case In theory - yes. However, UploadID it just a `uuid` string. It is simple to generate.
Trying the approach with first object id as an UploadID requires many code changes, extended control for this _zero_ object during uploading and completing.
I see many problems and no profit for changing it right now.

Comment (roman-khimov) [reply internal/neofs/tree.go:1343]: 1. Per S3 spec UploadID is just a string. It can be anything, including OID.
2. Client will use this UploadID for _every_ subsequent request which means you'll get this OID directly, you do not need to map UUID to first OID.
3. The first object would then be _always_ available, even if you're reuploading the first chunk.
4. There is nothing to control there, just create it for `CreateMultipartUpload`. Nothing you need to do with it afterwards (except for regular usage as first object OID).

Comment (smallhive) [reply internal/neofs/tree.go:1343]: Do we do this inside this PR or in a separate issue?

Comment (roman-khimov) [reply internal/neofs/tree.go:1343]: It's a part of v2 switch to me. It was specifically designed to allow for it.

File: api/layer/tree_service.go (hunk 2 above)
Comment (carpawell) [reply api/layer/tree_service.go:80]: > V1 has only previous link

yes, but if you change the first one, you have to change `previous` for every next file, cause every next part should be recalculated... but it wasn't like that before?

Comment (carpawell) [reply api/layer/tree_service.go:80]: also, i saw that the last part is always zero-length (in NeoFS terms), can we put the first one with an empty payload too just to fix the first child ID? it will not be reuploaded then ever. @roman-khimov

Comment (roman-khimov) [reply api/layer/tree_service.go:80]: https://github.com/nspcc-dev/neofs-s3-gw/pull/957#discussion_r1634639794

Comment (carpawell) [reply api/layer/tree_service.go:80]: whoops, missed that, i am totally "for" uploading the first object as an initial multipart uploading ID and fixing this important split part

File: internal/neofs/tree.go (hunk 1 above)
Comment (smallhive) [reply internal/neofs/tree.go:1335]: Updated

**Review by roman-khimov — COMMENTED** [review]
I've expected more red lines, but probably we're still too dependent on the tree service.

File: api/layer/multipart_upload.go [hunk 12]
```diff
@@ -417,6 +420,114 @@ func (n *layer) uploadPart(ctx context.Context, multipartInfo *data.MultipartInf
...
+	encInfo := FormEncryptionInfo(multipartInfo.Meta)
+	if err := p.Encryption.MatchObjectEncryption(encInfo); err != nil {
+		n.log.Warn("mismatched obj encryptionInfo", zap.Error(err))
+		return nil, s3errors.GetAPIError(s3errors.ErrInvalidEncryptionParameters)
+	}
+
+	var (
+		bktInfo         = p.Bkt
+		attributes      [][2]string
+		multipartHash   = sha256.New()
```
[the remark this answers is not in the document]
Comment (smallhive) [reply api/layer/multipart_upload.go:433]: It is correct. Unfortunately, I failed to avoid it. Even empty data must be hashed and marshaled to be available for the next parts. This required calculating a hash for the whole object and the next part uses the "hash state" from the previous one

File: api/handler/object_list.go [hunk 13]
```diff
@@ -277,7 +278,7 @@ func encodeListObjectVersionsToResponse(info *layer.ListObjectVersionsInfo, buck
 			Key:          ver.ObjectInfo.Name,
 			LastModified: ver.ObjectInfo.Created.UTC().Format(time.RFC3339),
 			Owner: Owner{
-				ID:          ver.ObjectInfo.Owner.String(),
+				ID:          hex.EncodeToString(ver.ObjectInfo.OwnerPublicKey.Bytes()),
```
Comment (roman-khimov) [inline api/handler/object_list.go:313]: StringCompressed?

File: api/layer/tree_service.go (hunk 2 above)
Comment (roman-khimov) [reply api/layer/tree_service.go:80]: So, do we still need this `number` parameter?

Comment (smallhive) [reply api/layer/tree_service.go:80]: Sure, we still have to take the previous part to take hashes from them. Just take the latest part isn't helpful for us

Comment (roman-khimov) [reply api/layer/tree_service.go:80]: But `GetLastPart` did exactly that without `number`, no?

Comment (smallhive) [reply api/layer/tree_service.go:80]: In common case of course, in reuploading case no
For instance: we have parts [1,2,3]. We are reuploading part 2. `GetLastPart` will return 3, which is not valid. GetPartByNumber will return 1 (because we take currentPartID - 1) - as the last stable point

**Review by roman-khimov — APPROVED** [review]

**Review by carpawell — APPROVED** [review]
This is (has been for 2 weeks) my first experience with our S3 review. Overall, looks working to me. However, I still do not understand why some changes are done _now_ and in the _replace v1 to v2_ commit. We have not changed the split noticeably, it is just now more convenient and more informative for SNs, but it is still an object chain, and still, the whole chain is backward-dependent, so it is strange.

Also, reuploading parts logic blows my mind, but it is not this PR's scope.
